relayer/aztec/relayer.go

```diff
@@ -0,0 +1,380 @@
+    }]`
+
+	parsedABI, err := abi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return false, fmt.Errorf("ABI parse error: %w", err)
+	}
+
+	data, err := parsedABI.Pack("parseAndVerifyVM", vaaBytes)
+	if err != nil {
+		return false, fmt.Errorf("ABI pack error: %w", err)
+	}
+
+	targetAddr := common.HexToAddress(targetContract)
+	msg := ethereum.CallMsg{
+		To:   &targetAddr,
+		Data: data,
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	result, err := c.client.CallContract(ctxWithTimeout, msg, nil)
+	if err != nil {
+		return false, fmt.Errorf("contract call failed: %w", err)
+	}
+
+	// Extract the "valid" boolean from the result
+	if len(result) >= 64 {
+		return result[63] == 1, nil // Check the last byte of the second 32-byte word
+	}
+
+	return false, fmt.Errorf("unexpected result format")
+}
+
+// Relayer coordinates processing VAAs from the spy service
+type Relayer struct {
+	spyClient    *SpyClient
+	evmClient    *EVMClient
+	config       Config
+	vaaProcessor func(*Relayer, *VAAData) error
+}
+
+// NewRelayer creates a new relayer instance
+func NewRelayer(config Config) (*Relayer, error) {
+	log.Println("[Relayer] Initializing...")
+
+	// Connect to the spy service
+	spyClient, err := NewSpyClient(config.SpyRPCHost)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create spy client: %w", err)
+	}
+
+	// Connect to the EVM chain
+	evmClient, err := NewEVMClient(config.DestRPCURL, config.PrivateKey)
+	if err != nil {
+		spyClient.Close()
+		return nil, fmt.Errorf("failed to create EVM client: %w", err)
+	}
+
+	relayer := &Relayer{
+		spyClient: spyClient,
+		evmClient: evmClient,
+		config:    config,
+	}
+
+	// Set default VAA processor
+	relayer.vaaProcessor = defaultVAAProcessor
+
+	return relayer, nil
+}
+
+// Close cleans up resources used by the relayer
+func (r *Relayer) Close() {
+	if r.spyClient != nil {
+		r.spyClient.Close()
+	}
+}
+
+// Start begins listening for VAAs and processing them
+func (r *Relayer) Start(ctx context.Context) error {
+	log.Printf("[Relayer] Starting with address: %s", r.evmClient.GetAddress().Hex())
+	log.Printf("[Relayer] Filtering for VAAs from chain %d to chain %d", r.config.SourceChainID, r.config.DestChainID)
+	log.Printf("[Relayer] Monitoring emitter address: %s", r.config.EmitterAddress)
+
+	// Subscribe to VAAs
+	stream, err := r.spyClient.SubscribeSignedVAA(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to VAA stream: %w", err)
+	}
+
+	log.Println("[Relayer] Listening for VAAs...")
+	for {
+		select {
+		case <-ctx.Done():
+			log.Println("[Relayer] Shutting down relayer.")
+			return nil
+		default:
+			// Receive the next VAA
+			resp, err := stream.Recv()
+			if err != nil {
+				log.Printf("[Relayer] Stream error: %v. Retrying in 5s...", err)
+				time.Sleep(5 * time.Second)
+				stream, _ = r.spyClient.SubscribeSignedVAA(ctx)
+				continue
+			}
+
+			// Process the VAA
+			go r.processVAA(ctx, resp.VaaBytes)
```

Comment:
We probably need a wait group. We want all the goroutines to finish before this function exits. It appears there is a race condition between this line and the `<-ctx.Done()` case.